app/javascript/lib/plot.js

```diff
@@ -9,62 +10,259 @@ const colorBrewerList = [
   return colorBrewerList[index % 27]
 }
 
 // default title font settings for axis titles in plotly
-export const titleFont = {
+PlotUtils.titleFont = {
   family: 'Helvetica Neue',
   size: 16,
   color: '#333'
 }
 
 // default label font settings for colorbar titles in plotly
-export const labelFont = {
+PlotUtils.labelFont = {
   family: 'Helvetica Neue',
   size: 12,
   color: '#333'
 }
 
-export const lineColor = 'rgb(40, 40, 40)'
+PlotUtils.lineColor = 'rgb(40, 40, 40)'
 
-/**
- * Wrapper for Plotly.newPlot, to enable tests
- *
- * Having this function in a separate module is needed for mocking in related
- * test (study-violin-plot.test.js), due to (1) and buggy workaround (2).
- *
- * 1) SVG path getTotalLength method is undefined in jsdom library used by Jest
- *    Details: https://github.com/jsdom/jsdom/issues/1330
+/** returns an empty trace with arrays initialized based on expectedLength */
+function emptyTrace(expectedLength, hasZvalues, hasExpression) {
+  const trace = {
+    x: new Array(expectedLength),
+    y: new Array(expectedLength),
+    annotations: new Array(expectedLength),
+    cells: new Array(expectedLength),
+    newLength: 0 // used to track the post-filter length -- NOT a plotly property
+  }
+  if (hasZvalues) {
+    trace.z = new Array(expectedLength)
+  }
+  if (hasExpression) {
+    trace.expression = new Array(expectedLength)
+  }
+  return trace
+}
+
+/** takes a plotly trace argument and filters out cells based on params.  will return the original object,
+ * but with data arrays filtered as appropriate
+ * @param hiddenTraces {String[]} array of label names to filter out
+ * @param activeTraceLabel if specified, the traces will be sorted such that the activeTrace is plotted on top
+ * @param groupByAnnotation {Boolean} whether to assemble separate traces for each label
  *
- * 2) jest.mock() does not work when module name has ".js" in it
- *    Details: https://github.com/facebook/jest/issues/6420
+ * For performance
  */
-export function plot(graphElementId, data, layout) {
-  Plotly.newPlot(
-    graphElementId,
-    data,
-    layout
-  )
+PlotUtils.filterTrace = function({
+  trace, hiddenTraces=[], groupByAnnotation=false,
+  activeTraceLabel, expressionFilter, expressionData
+}) {
+  const isHidingByLabel = hiddenTraces && hiddenTraces.length
+  const isFilteringByExpression = expressionFilter && expressionData &&
+    (expressionFilter[0] !== 0 || expressionFilter[1] !== 1)
+  const hasZvalues = !!trace.z
+  const hasExpression = !!trace.expression
+  const oldLength = trace.x.length
+  // if grouping by annotation, traceMap is a hash of annotation names to traces
+  // otherwise, traceMap will just have a single 'all' trace
+  const traceMap = {}
+  const estTraceLength = groupByAnnotation ? Math.round(trace.x.length / 10) : trace.x.length
+  let rawCountsByLabel = { main: trace.x.length } // maintain a list of all cell counts by label for the legend
```

Comment:
`countsByLabel` is used later in the function for the post-filtered counts.  I appreciate the confusion since 'rawCounts' is a loaded term.  I'll switch to "unfilteredCountsByLabel"